### PR TITLE
Add selection border to layer widgets

### PR DESCRIPTION
**Pull request by sofroniewn**
# What does this PR do? Why are doing this? References?
This PR adds the ability to select layer widgets by clicking on their `groupbox`. Clicking changes the border of the box to a focus color. It also changes the `layer.selected` property to `True`. Multiple layer widgets can be selected by holding and clicking with the shift key down. This PR also removes colormap changing (to be readded at a later time, in some form of popup like box) and adds the ability to edit a layers name by typing in the name text edit box. This PR also adds an eye on / off icon to the visibility toggle.

When new layers are created they do not automatically become selected, nor do they affect the selection of other layers. This could be changed if desired.

Right now there is no concept of select all layers between layer A and layer B. Something this is done using modifier keys different from those used to select only A and B. This could be implemented in the future too.

I think this PR is ready to be merged, and integrated with the clims / gamma sliders that @AhmetCanSolak is working on. On thing that is unclear to be in that linking will be how to deal with the case where multiple layers are selected or no layers are selected, but that is for that PR to address.

## Type of change
[Please delete options that are not relevant.]
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## How Has This Been Tested?
[Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce.]
- [x] TestCase Ran `examples/layers.ipynb` and interacted with gui.

## Final Checklist:
- [x] My PR is the possible minimum work for the desired functionality
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works

### CODE REVIEW

**Review by kne42 — COMMENTED** [review]
Great job, it looks very nice!

A few general comments:
- you should remove `.DS_Store`
- you should include the `icons` folder in the `MANIFEST.in`
- I don't think that the `eye_off` icon is super clear. IMHO would be better if you put a slash or something like that through it.
- Regarding selection semantics, I have opened #55 for discussion.
- I think that it should be more clear on when the layer name is saved. One design that I really like is a little checkmark button appearing to the side and you can either hit enter or click it to save your changes and dismiss this button from view. This can be done in a later PR tho

**Comment by sofroniewn** [comment]
> you should remove .DS_Store

Done

> you should include the icons folder in the MANIFEST.in

Done

> I don't think that the eye_off icon is super clear. IMHO would be better if you put a slash or something like that through it.

On your suggestion I tried this, but it still didn't look so clear, so instead I replaced the `off` state icon with the normal empty checkbox - some other similar applications do this too. It's easy for me to change again though if people want.

> Regarding selection semantics, I have opened #55 for discussion.

I have tried to improve this copying the suggestions there. Give it a test to see if it works as expected. There are some weird potential corner cases, but I tried to deal with them. One thing I'll note here is that it is currently impossible to have NO layers selected. I can change this if desired. Another thing I'll note is that when a new layer gets added it is automatically selected, but it doesn't unselect the previous layers. I think this behaviour makes more sense, but I havn't implemented it yet.

> I think that it should be more clear on when the layer name is saved. One design that I really like is a little checkmark button appearing to the side and you can either hit enter or click it to save your changes and dismiss this button from view. This can be done in a later PR tho

I'm not so sure about this one, now the update function gets called when editing is done. It's a pretty small function, just changing the name property, so I'm not really worried about it getting called to frequently. But happy to hear what others think too.

I've made a couple other minor changes. One is to reorder the visibility / opacity / and name widgets, to be in an order I think is cleaner. Another change is that new layers now get added to the top of the layer list and push the others down. I think this is also cleaner as newer things are more easily accessible, but happy to discuss too.

From my end I think the PR is ready to go, but please check over.

File: gui/elements/qt/_imageLayer.py
Comment (kne42) [inline gui/elements/qt/_imageLayer.py:8]: Why did this get disabled anyways?

File: .gitignore
Comment (kne42) [inline .gitignore]: IMO `.DS_Store` and other OS-specific files do not belong here. You should look into adding a global `.gitignore` via:
```
# ~/.gitconfig
[core]
    excludesfile = /path/to/my/global/gitignore
```

You can find a full example [here](https://github.com/kne42/dotfiles/tree/master/git).

**Review by kne42 — APPROVED** [review]
> I'm not so sure about this one, now the update function gets called when editing is done. It's a pretty small function, just changing the name property, so I'm not really worried about it getting called to frequently.

Oh no, I meant making it more clear that the editing was over, not propagating the changes on the fly. For example, if I have edited the name and I press enter, I don't see the text box being deselected even though the change has gone through.

File: gui/elements/qt/_imageLayer.py
Comment (sofroniewn) [reply gui/elements/qt/_imageLayer.py:8]: I removed the cmap selection from the main layer widget. It will now be accessible by double clicking on the layer widget along with other custom layer specific properties. I havn't added that back in yet, but it will be my next PR after this one is merged.

File: .gitignore
Comment (sofroniewn) [reply .gitignore]: got it - removing change, and replaced with global git ignore on my computer

**Comment by sofroniewn** [comment]
I have made the unselection changes requested in #55 and updated the PR

File: gui/elements/qt/_imageLayer.py
Comment (jni) [reply gui/elements/qt/_imageLayer.py:8]: Will double-clicking be replaced by a sandwich menu at some point?

Comment (sofroniewn) [reply gui/elements/qt/_imageLayer.py:8]: I'm open to either, I have created a new issue to discuss this #59. I think it will be in a new PR regardless, currently neither are implemented. It's one of the next things on my list though after this one get's merged.

**Comment by jni** [comment]
Ok I'm officially in the habit of trying out PRs now. =)

Not that I think any of these comments should hold up the PR, but:
- when did we change the default colour map to "hot" and can we change it to "inferno" or "magma"? Or, as I mentioned, something created with the cmap creator that enabled those colormaps.
- I think the layer title should be in a white box while editing. Currently it's very unclear that I can edit it. The only reason I even tried it is that I remembered the discussion here.
- I wonder if it's possible to automagically grab the variable name being passed to add_layer so that it is set to that from the start. It might be worth seeing how Spyder populate their workspace viewer.
- there was some discussion of dragging and dropping to reorder layers. That's not supposed to be working yet, right? Cos it's not for me.

Finally, can someone please set "Ctrl/Cmd-W" to close the viewer window? =P

**Comment by jni** [comment]
Oh btw, the most important thing: the selection indeed works as I would expect! =)

**Comment by sofroniewn** [comment]
@jni great! glad the layer reordering worked as expected - can you approve the PR ;-)

As to your other comments

> when did we change the default colour map to "hot" and can we change it to "inferno" or "magma"? Or, as I mentioned, something created with the cmap creator that enabled those colormaps.

I think the plan here once we have a preferences box and file enable will be to put the default here so users can set it

> I think the layer title should be in a white box while editing. Currently it's very unclear that I can edit it. The only reason I even tried it is that I remembered the discussion here.

Ok I can make this change, will wait till the next PR if ok.

> I wonder if it's possible to automagically grab the variable name being passed to add_layer so that it is set to that from the start. It might be worth seeing how Spyder populate their workspace viewer.

Yes would like to do something like this, or at least have the default auto increment so new layers have higher numbers.  

> there was some discussion of dragging and dropping to reorder layers. That's not supposed to be working yet, right? Cos it's not for me.

Doesn't work yet, will be the next PR after this gets merged. Have been playing around with that concept in a little demo and think i'll be able to make a first pass at it pretty quickly

> Finally, can someone please set "Ctrl/Cmd-W" to close the viewer window? =P

Keyboard shortcuts should be another PR soon too

**Comment by kne42** [comment]
@jni
1) I did that a while ago because I figured it was the closest to those. They don’t exist yet though.
2) Agreed, this was one thing I found ambiguous as well.
3) Yes we can accomplish this by inspecting the call stack. However, parsing this information is not trivial since statements may be nested. For example, `imshow` calls `add_image`, what if another function calls that, etc. Do we base the name off of what variable it is assigned to or somehow using the variables passed into the function as parameters? Regardless, I think somewhere down the line, we’ll want to bring in someone with experience parsing syntax (probably a programming languages specialist) since we have quite a few features now in mind that would require this. The list so far is: autogenerating layer names from call context, checking when a variable is assigned to (for IPython updating), and determining all type hints in a given context (for more magic methods). Needless to say this is certainly not a priority, but it could be something we look at in a post-1.0 world.

**Review by AhmetCanSolak — APPROVED** [review]

**Pull request merged**
